src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp

```diff
@@ -90,4 +97,175 @@ size_t overlap_num_points(const Index<Dim>& volume_extents,
 double overlap_width(size_t overlap_extent,
                      const DataVector& collocation_points) noexcept;
 
+/// Iterate over grid points in a region that extends partially into the volume
+class OverlapIterator {
+ public:
+  template <size_t Dim>
+  OverlapIterator(const Index<Dim>& volume_extents, size_t overlap_extent,
+                  const Direction<Dim>& direction) noexcept;
+
+  explicit operator bool() const noexcept;
+
+  OverlapIterator& operator++();
+
+  /// Offset into a DataVector that holds full volume data
+  size_t volume_offset() const noexcept;
+
+  /// Offset into a DataVector that holds data only on the overlap region
+  size_t overlap_offset() const noexcept;
+
+  void reset() noexcept;
+
+ private:
+  size_t size_ = std::numeric_limits<size_t>::max();
+  size_t num_slices_ = std::numeric_limits<size_t>::max();
+  size_t stride_ = std::numeric_limits<size_t>::max();
+  size_t stride_count_ = std::numeric_limits<size_t>::max();
+  size_t jump_ = std::numeric_limits<size_t>::max();
+  size_t initial_offset_ = std::numeric_limits<size_t>::max();
+  size_t volume_offset_ = std::numeric_limits<size_t>::max();
+  size_t overlap_offset_ = std::numeric_limits<size_t>::max();
+};
+
+// @{
+/// The part of the tensor data that lies within the overlap region
+template <size_t Dim, typename DataType, typename... TensorStructure>
+void data_on_overlap(const gsl::not_null<Tensor<DataType, TensorStructure...>*>
+                         restricted_tensor,
+                     const Tensor<DataType, TensorStructure...>& tensor,
+                     const Index<Dim>& volume_extents,
+                     const size_t overlap_extent,
+                     const Direction<Dim>& direction) noexcept {
+  for (OverlapIterator overlap_iterator{volume_extents, overlap_extent,
+                                        direction};
+       overlap_iterator; ++overlap_iterator) {
+    for (decltype(auto) overlap_and_volume_tensor_components :
```

Comment:
tensors can just be indexed directly:
```
for(size_t i = 0; i < tensor.size(); ++i) {
// do something with tensor[i]
}
```
or am I misunderstanding what you're doing here?

Reply:
This is just a loop over the tensor components. Looping over the tensor size would also work. This code is adapted from `Tensor/Slice.hpp` which does the loop over boost-combine. If there's a reason to prefer one approach over the other I'd be happy to change it.

Reply:
I have the inclination that the iteration over the tensor size is slightly preferable. Because tensors are `std::array`s under the hood, the direct indexing should be both faster (but only a tiny gain) and more succinct code.  Besides, it means that this code doesn't need to depend on boost, which is also a small gain.